### ISSUE TITLE
Corpus to DFM conversion doesn't work/takes too much time

### ISSUE THREAD

**Issue by Dom-320** [quanteda/quanteda#1831]
I am fairly new to Quanteda and having issues when converting corpus to DFM - the conversion seems to be taking extremely long. (I couldn't even measure it with system. time). 
I am trying to convert a corpus made from a dataset that originally had around 300MB and 2367198 obs. Could anyone explain to me why is this taking so long or how could I make it work?

corpus_dfm <- dfm(corpus)

Thanks

**Comment by kbenoit** [comment]
Please show us the output of
- `sessionInfo()`
- `ndoc(corpus)`
- `dfm(corpus, verbose = TRUE`

We cannot tell you much without more details about what you mean by "extremely long" or without knowing some details about your system. But if you have 2.37 million documents, then yes that's a big process to tokenize and tabulate those documents into a dfm. I'm assuming that the texts are all in the same language?

**Comment by Dom-320** [comment]
Thank you. Yes, the texts are all in English, it's a dataset of tweets. I thought maybe converting the original dataset into a data.table would help but that does not seem to make it any faster either.

sessionInfo() and ndoc(corpus) outputs are below,  however, dfm(corpus, verbose = TRUE) is taking 5+ minutes plus by now without reaching an output.




Session info:
R version 3.6.1 (2019-07-05)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Sierra 10.12.6

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

Random number generation:
 RNG:     Mersenne-Twister 
 Normal:  Inversion 
 Sample:  Rounding 
 
locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] quanteda_1.5.1    data.table_1.12.8

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.2         rstudioapi_0.10    magrittr_1.5       stopwords_1.0     
 [5] tidyselect_0.2.5   munsell_0.5.0      colorspace_1.4-1   lattice_0.20-38   
 [9] R6_2.4.0           rlang_0.4.1        fastmatch_1.1-0    stringr_1.4.0     
[13] dplyr_0.8.3        tools_3.6.1        grid_3.6.1         gtable_0.3.0      
[17] spacyr_1.2         RcppParallel_4.4.4 lazyeval_0.2.2     assertthat_0.2.1  
[21] tibble_2.1.3       crayon_1.3.4       Matrix_1.2-17      purrr_0.3.3       
[25] ggplot2_3.2.1      curl_4.2           glue_1.3.1         stringi_1.4.3     
[29] compiler_3.6.1     pillar_1.4.2       scales_1.0.0       lubridate_1.7.4   
[33] pkgconfig_2.0.3   

ndcoc:
[1] 2367198

**Comment by kbenoit** [comment]
OK thanks. 5 minutes may not be a lot given the size here. Try this following (assuming your corpus is named `corpus`):

```r
tokens(corpus, what = "fastestword", verbose = TRUE) %>%
    dfm(verbose = TRUE)
```
Let it run, and upload the output.
tokens(

**Comment by Dom-320** [comment]
Output following all the blocks:

...total elapsed:  1194 seconds.
Finished tokenizing and cleaning 2,367,198 texts.
Creating a dfm from a tokens input...
   ... lowercasing
   ... found 2,367,198 documents, 1,899,987 features
   ... created a 2,367,198 x 1,899,987 sparse dfm
   ... complete. 
Elapsed time: 208 seconds.
Document-feature matrix of: 2,367,198 documents, 1,899,987 features (>99.99% sparse).

**Comment by Dom-320** [comment]
Does that mean the conversion to dfm would take 1194 seconds?

**Comment by koheiw** [comment]
Please, don't forget to set `quanteda_options(threads = 8)` or more. Try again using the Github master. We are developing v2.0 and there is no plan to fix v1.5.x even if there are problems.

**Comment by kbenoit** [comment]
On your system, it's taking 20 mins to tokenize this monster corpus, but only 3.5 mins to create the dfm from it. As you can see, this is a mostly empty dfm (the "(>99.99% sparse)" message).

So your best bet is to 
a) create the tokens object, save it (1194 secs);
b) create the dfm from the tokens object (208 seconds);
c) save the dfm;
d) trim the dfm using `dfm_trim()`, save that, and work with that object. Trying to do anything with a nearly empty dfm of that size is going to be pretty difficult, unless you trim the feature set.

Given that these are short documents (Tweets) it's also a good idea to think about grouping documents by date, user, etc to create fewer, larger documents.
